File: src/main/java/eu/mjindra/utils/properties/Money.java
package eu.mjindra.utils.properties;

import eu.mjindra.utils.units.Base;
import eu.mjindra.utils.units.Coin;
import org.apache.commons.math3.util.Precision;

/**
 * Monnnneeeeeyyyy.
 *
 * @author Martin Jindra
 * @version 02.06.2023
 */
public class Money extends Unit {

    public Money(float amount, Coin coin) {
        super(amount, coin);
    }

    /**
     * Convert the value of one unit to another unit.
     *
     * @param to different unit
     * @return converted value
     */
    @Override
    public Money convert(Base to) {
        Coin c = (Coin) to;
        float f = super.getAmount() / ((Coin) super.getUnit()).val;
        return new Money(Precision.round(f * c.val, 4), c);
    }
}

File: lib/dnd/properties/money.dart
import 'package:helios/dnd/coins.dart';
import 'package:helios/dnd/quantity/unit.dart';
import 'package:helios/dnd/quantity/value.dart';

/// Monnnneeeeeyyyy.
class Money extends Value {
  Money(double amount, Coin coin) : super(amount, coin);

  /// Convert the value of one unit to another unit.
  @override
  Money convert(Unit to) {
    Coin c = to as Coin;
    double d = super.amount * (super.unit as Coin).value;
    return Money(round(d / c.value), c);
  }
}
